internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -469,53 +469,35 @@ private[inc] abstract class IncrementalCommon(
       invalidateTransitively: Boolean,
       isScalaClass: String => Boolean
   ): Set[String] = {
+    val initial = changes.allModified.toSet
+    val dependsOnClass = findClassDependencies(_, relations)
     val firstClassInvalidation: Set[String] = {
-      if (invalidateTransitively) {
+      val invalidated = if (invalidateTransitively) {
         // Invalidate by brute force (normally happens when we've done more than 3 incremental runs)
-        val dependsOnClass = relations.memberRef.internal.reverse _
-        transitiveDependencies(dependsOnClass, changes.allModified.toSet)
+        IncrementalCommon.transitiveDeps(initial, log)(dependsOnClass)
       } else {
-        includeTransitiveInitialInvalidations(
-          changes.allModified.toSet,
-          changes.apiChanges.flatMap(invalidateClassesInternally(relations, _, isScalaClass)).toSet,
-          findClassDependencies(_, relations)
-        )
+        changes.apiChanges.flatMap(invalidateClassesInternally(relations, _, isScalaClass)).toSet
       }
+      val included = includeTransitiveInitialInvalidations(initial, invalidated, dependsOnClass)
+      log.debug("Final step, transitive dependencies:\n\t" + included)
+      included
     }
 
     // Invalidate classes linked with a class file that is produced by more than one source file
     val secondClassInvalidation = IncrementalCommon.invalidateNamesProducingSameClassFile(relations)
     if (secondClassInvalidation.nonEmpty)
       log.debug(s"Invalidated due to generated class file collision: ${secondClassInvalidation}")
 
-    val newInvalidations = (firstClassInvalidation -- recompiledClasses) ++ secondClassInvalidation
-    if (newInvalidations.isEmpty) {
+    val allInvalidations = (firstClassInvalidation -- recompiledClasses) ++ secondClassInvalidation
+    if (allInvalidations.isEmpty) {
       log.debug("No classes were invalidated.")
       Set.empty
     } else {
-      val allInvalidatedClasses: Set[String] = firstClassInvalidation ++ secondClassInvalidation
-      log.debug(s"Invalidated classes: ${allInvalidatedClasses.mkString(", ")}")
-      allInvalidatedClasses
+      log.debug(s"Invalidated classes: ${allInvalidations.mkString(", ")}")
+      allInvalidations
```

Comment:
Maybe the bug is here?
In the old code, it's checking for the diff, but returns `firstClassInvalidation ++ secondClassInvalidation`

Reply:
Maybe.  Why does this have two different equations?

Reply:
It's checking if there's anything new that are being invalidated, but it returns both new and old invalidation together.

Reply:
But the regression proves that the "old" invalidations aren't ignorable, no?  Therefore I think they shouldn't be ignored in that diff/if (i.e. do the opposite alignment that I went for)

Reply:
I think the regression indicates that you can ignore old when there's no new invalidation, but both old and new invalidations must be compiled together to avoid circular situation in some cases.

Reply:
How does it indicate the former?

Reply:
The old-and-working code says:

```scala
    val newInvalidations = (firstClassInvalidation -- recompiledClasses) ++ secondClassInvalidation
    if (newInvalidations.isEmpty) {
      log.debug("No classes were invalidated.")
      Set.empty
```

Reply:
OK, I don't think the regression indicates that you can ignore the old invalidations.

The question is will `firstClassInvalidation ++ secondClassInvalidation` ever be non-empty due to `recompiledClasses` alone?  That would suggests recompiledClasses invalidates itself, which sounds like an invariant that isn't going to happen.  So we can just drop the whole `-- recompiledClasses` business and align like that.